src/main/java/org/jvnet/hudson/test/RealJenkinsRule.java

```diff
@@ -608,25 +618,42 @@ private static boolean supportsPortFileName(String war) throws IOException {
 
     private static int readPort(File portFile) throws IOException {
         String s = FileUtils.readFileToString(portFile, StandardCharsets.UTF_8);
+        if (StringUtils.isEmpty(s)) {
+            LOGGER.log(Level.WARNING, () -> String.format("PortFile: %s exists, but value is still not written", portFile.getAbsolutePath()));
+            return 0;
+        }
         try {
             return Integer.parseInt(s);
         } catch (NumberFormatException e) {
             throw new AssertionError("Unable to parse port from " + s + ". Jenkins did not start.");
         }
     }
 
+    /**
+     * Stops Jenkins and releases any system resources associated
+     * with it. If Jenkins is already stopped then invoking this
+     * method has no effect.
+     */
     public void stopJenkins() throws Throwable {
-        endpoint("exit").openStream().close();
-        if (!proc.waitFor(60, TimeUnit.SECONDS) ) {
-            System.err.println("Jenkins failed to stop within 60 seconds, attempting to kill the Jenkins process");
-            proc.destroyForcibly();
-            throw new AssertionError("Jenkins failed to terminate within 60 seconds");
+        if (!stopping.get()) {
+            stopping.set(true);
+            endpoint("exit").openStream().close();
         }
-        int exitValue = proc.exitValue();
-        if (exitValue != 0) {
-            throw new AssertionError("nonzero exit code: " + exitValue);
+        if (proc != null) {
+            try {
+                if (!proc.waitFor(60, TimeUnit.SECONDS) ) {
+                    System.err.println("Jenkins failed to stop within 60 seconds, attempting to kill the Jenkins process");
+                    proc.destroyForcibly();
+                    throw new AssertionError("Jenkins failed to terminate within 60 seconds");
+                }
+                int exitValue = proc.exitValue();
+                if (exitValue != 0) {
+                    throw new AssertionError("nonzero exit code: " + exitValue);
+                }
+            } finally {
+                proc = null;
```

Comment:
I suspect you could dispense with `stopping` altogether and just do something like

```java
if (proc != null) {
  Proc _proc = proc;
  proc = null;
  // …now terminate _proc as before
}
```

Reply:
I would say it should work, do you prefer this approach over atomic synchronisation using `stopping`?

Reply:
I like this idea and applied changes in [2430d48](https://github.com/jenkinsci/jenkins-test-harness/pull/465/commits/2430d48729c5e81f362fba229c6f20bd0e316bf3)